server/schedule/operator_controller.go

```diff
@@ -93,6 +93,26 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
 		timeout := op.IsTimeout()
 		if step := op.Check(region); step != nil && !timeout {
 			operatorCounter.WithLabelValues(op.Desc(), "check").Inc()
+
+			// When the source is heartbeat, the region may have a newer
+			// epoch than the region that the operator holds. In this case,
+			// the operator is stale, and will not be executed even we would
+			// have sent it to TiKV servers. Here, we just cancle it.
+			origin := op.RegionEpoch()
+			latest := region.GetRegionEpoch()
+			changes := (latest.GetVersion() - origin.GetVersion()) +
+				(latest.GetConfVer() - origin.GetConfVer())
+			if source == DispatchFromHeartBeat &&
+				changes > uint64(op.CurrentStep()) {
```

Comment:
It seems to be insufficient to compare version changes to step because there are some steps does not change versions. For example, steps are '[transferLeader, removePeer]', tikv first executes `transferLeader` with no version change, then executes an unexpected `removePeer` which changes version by 1. The region will pass the check here and execute another `removePeer`.

Reply:
I though the epoch will be changed for every operator step. If not, we must record if the epoch would be changed for every step, otherwise, we can not guarantee the atomicity.

Reply:
Yes. It's kind tricky to resolve the problem. I wonder if we can add the expected epoch for each step, when the epoch does not match expected one, we abort the whole operator.

Reply:
Maybe we can sync raft term from tikv to PD to resolve this problem.

Reply:
Given that we should add a method for every `OpStep`, the term is unnecessary in fact.

I added a method called `ConsumeConfVer`, which returns the number of conf-version should be consumed if this step was executed. A step does not change the conf-version just return 0.